numpy/core/shape_base.py

```diff
@@ -423,32 +423,74 @@ def format_index(index):
         return parent_index, _nx.ndim(arrays)
 
 
-def _block(arrays, max_depth, result_ndim):
+def _shape_ndim(array, ndim):
+    """Prepends leadings ones to the shape of the array up to `ndim`."""
+    return (1, ) * (ndim - array.ndim) + array.shape
+
+
+def _assignment_recursion(result, arrays, list_ndim, start_indicies,
+                          depth=0):
+    if depth < list_ndim:
+        concat_depth = depth + result.ndim - list_ndim
+        for arr in arrays:
+            shape = _assignment_recursion(result, arr, list_ndim,
+                                          start_indicies, depth=depth+1)
+            start_indicies[concat_depth] += shape[concat_depth]
+        # lists are mutable, so we have to reset the start_indicies list
+        start_indicies[concat_depth:] = (0, ) * (result.ndim - concat_depth)
+        return shape
+    else:
+        # We've bottom out, assign the array
+        shape = _shape_ndim(arrays, result.ndim)
+        slices = tuple(slice(start, start+size)
+                       for start, size in zip(start_indicies, shape))
+        result[slices] = arrays
+        return shape
+
+
+def _shape_recursion(arrays, list_ndim, result_ndim, depth=0):
+    """Internal implementation that gets the shape of the final array.
     """
-    Internal implementation of block. `arrays` is the argument passed to
-    block. `max_depth` is the depth of nested lists within `arrays` and
-    `result_ndim` is the greatest of the dimensions of the arrays in
-    `arrays` and the depth of the lists in `arrays` (see block docstring
-    for details).
+    if depth < list_ndim:
+        shapes = [_shape_recursion(arr, list_ndim, result_ndim, depth+1)
+                  for arr in arrays]
+        concat_depth = depth + result_ndim - list_ndim
+        for shape in shapes[1:]:
+            if (shape[:concat_depth] != shapes[0][:concat_depth] or
+                    shape[concat_depth+1:] != shapes[0][concat_depth+1:]):
+                raise ValueError('Mismatched array shapes in block.')
+        shape_on_dim = sum(shape[concat_depth] for shape in shapes)
+        # Take a shape, any shape
+        shape = shapes[0]
+        shape = shape[:concat_depth] + (shape_on_dim,) + shape[concat_depth+1:]
```

Comment:
Helper function for lines 457 to 465: `result_shape = _concatenate_shape((shape1, shape2, ...), axis=axis)`